dataset/batch_image.py

```diff
@@ -1,525 +1,717 @@
-            origin = 0, 0
-        elif origin == CROP_CENTER:
-            origin = (image.width - shape[0]) // 2, (image.height - shape[1]) // 2
-        origin_x, origin_y = origin
-        shape = shape if shape is not None else (image.width - origin_x, image.height - origin_y)
-        box = origin_x, origin_y, origin_x + shape[0], origin_y + shape[1]
-        new_image = image.crop(box)
-        return new_image
+    def _invert_(self, image, channels='all'):
+        """ Invert channels
 
-    @inbatch_parallel('indices', post='assemble')
-    def _crop(self, ix, components='images', origin=None, shape=None):
-        """ Crop all images """
-        image = self.get(ix, components)
-        new_image = self._crop_image(image, origin, shape)
-        return new_image
+        Parameters
+        ----------
+        components : str
+            component to get an image from
+        p : float
+            probability of applying the transforms
+        channels : int, sequence
+            channels indices to invert.
 
-    @inbatch_parallel('indices', post='assemble')
-    def _random_crop(self, ix, components='images', shape=None):
-        """ Crop all images with a given shape and a random origin """
-        image = self.get(ix, components)
-        origin_x = np.random.randint(0, image.width - shape[0])
-        origin_y = np.random.randint(0, image.height - shape[1])
-        new_image = self._crop_image(image, (origin_x, origin_y), shape)
-        return new_image
+        Returns
+        -------
+        np.ndarray : inverted image
+        """
 
-    @action
-    @inbatch_parallel('indices', post='assemble')
-    def fliplr(self, ix, components='images'):
-        """ Flip image horizontaly (left / right) """
-        image = self.get(ix, components)
-        return image.transpose(PIL.Image.FLIP_LEFT_RIGHT)
+        if channels == 'all':
+            channels = list(range(image.shape[-1]))
+        inv_multiplier = 255 if np.issubdtype(image.dtype, np.integer) else 1.
+        image[..., channels] = inv_multiplier - image[..., channels]
+        return image
 
-    @action
-    @inbatch_parallel('indices', post='assemble')
-    def flipud(self, ix, components='images'):
-        """ Flip image verticaly (up / down) """
-        image = self.get(ix, components)
-        return image.transpose(PIL.Image.FLIP_TOP_BOTTOM)
+
+
+    def _salt_(self, image, p_pixel=.015, salt=255):
+        """ set random pixel on image to the givan value
+
+        every pixel will be set to ``salt`` value with probability ``p_pixel``
+
+        Parameters
+        ----------
+        components : str
+            component to get an image from
+        p : float
+            probability of applying the transforms
+        image : np.ndarray
+            image to flavour with species
+        p_pixel : float
+            probability of salting a pixel
+        salt : float, int, tuple
+            salt's value
+
```

Comment:
`size` is also needed
coarse noise is extremely helpful
it can also be a distribution